osutil/disks/gpt.go

```diff
@@ -0,0 +1,155 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package disks
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"hash/crc32"
+	"os"
+)
+
+type GPTLBA uint64
+type GPTGUID [16]byte
+
+// https://en.wikipedia.org/wiki/GUID_Partition_Table#Partition_table_header_(LBA_1)
+type GPTHeader struct {
+	Signature      [8]byte
+	Revision       uint32
+	HeaderSize     uint32
+	HeaderCRC      uint32
+	Reserved       uint32
+	CurrentLBA     GPTLBA
+	AlternateLBA   GPTLBA
+	FirstUsableLBA GPTLBA
+	LastUsableLBA  GPTLBA
+	DiskGUID       [16]byte
+	EntriesLBA     GPTLBA
+	NEntries       uint32
+	EntrySize      uint32
+	EntriesCRC     uint32
+}
+
+func verifyHeader(rawHeader []byte, header GPTHeader) error {
+	if !bytes.Equal(header.Signature[:], []byte("EFI PART")) {
+		return fmt.Errorf("GPT Header does not start with the magic string")
+	}
+
+	if header.Revision != 1<<16 {
+		return fmt.Errorf("GPT header revision is not 1.0")
+	}
+
+	if int(header.HeaderSize) < binary.Size(header) {
+		return fmt.Errorf("GPT header size is smaller than the minimum valid size")
+	}
+
+	if int(header.HeaderSize) > len(rawHeader) {
+		return fmt.Errorf("GPT header size is larger than the maximum supported size")
+	}
+
+	for i := 0; i < 4; i++ {
+		rawHeader[8+4+4+i] = 0
+	}
+	{
+		crc := crc32.ChecksumIEEE(rawHeader[:header.HeaderSize])
+		if crc != header.HeaderCRC {
+			return fmt.Errorf("GPT header CRC32 checksum failed: %v != %v", crc, header.HeaderCRC)
+		}
+	}
+
+	return nil
+}
+
+func LoadGPTHeader(devfd *os.File) (GPTHeader, error) {
+	var header GPTHeader
+	rawHeader := make([]byte, 512)
+	read, err := devfd.Read(rawHeader)
+	if err != nil {
+		return GPTHeader{}, err
+	}
+
+	if read < binary.Size(header) {
+		return GPTHeader{}, fmt.Errorf("Not enough data was read")
+	}
+
+	buf := bytes.NewReader(rawHeader[:read])
+	binary.Read(buf, binary.LittleEndian, &header)
+
+	if err := verifyHeader(rawHeader[:read], header); err != nil {
+		return GPTHeader{}, err
+	}
+
+	return header, nil
+}
+
+func ReadGPTHeader(device string) (GPTHeader, error) {
+	devfd, err := os.Open(device)
+	if err != nil {
+		return GPTHeader{}, err
+	}
+	defer devfd.Close()
+
+	if _, err := devfd.Seek(512, 0); err != nil {
+		return GPTHeader{}, err
+	}
+
+	header, main_err := LoadGPTHeader(devfd)
+	if main_err != nil {
+		// Read the backup header
+		_, err := devfd.Seek(-512, 2)
+		if err != nil {
+			return GPTHeader{}, main_err
+		}
+
+		header, err = LoadGPTHeader(devfd)
+		if err != nil {
+			return GPTHeader{}, main_err
+		}
+	}
+
+	return header, nil
+}
+
+func CalculateLastUsableLBA(device string) (uint64, error) {
+	header, err := ReadGPTHeader(device)
+	if err != nil {
+		return 0, err
+	}
+	sectors, err := blockDeviceSizeInSectors(device)
+	if err != nil {
+		return 0, err
+	}
+	tableSize := uint64(header.NEntries) * uint64(header.EntrySize)
+	// Rounded up division for number of sectors
+	tableSizeInSectors := (tableSize + 511) / 512
```

Comment:
I'm not sure that we can hardcode 512 as the sector size here; we do have a `blockDeviceSectorSize()` function, so it would be better to use that for consistency.

Reply:
We are talking about *logical* blocks. So it is always 512. `blockDeviceSectorSize` could give us something that is not the same as logical blocks.

Reply:
Ah, I see. And `blockDeviceSizeInSectors()` also reports logical sectors, so this is indeed correct.